### PR TITLE
tune the "Wal segment size" with a flag

### PR DESCRIPTION
**Pull request by glutamatt**
Add WALSegmentSize as an option, and the corresponding flag "storage.tsdb.wal-segment-size" to tune the max size of wal segment files.

The addressed base problem is to reduce the disk space used by wal segment files : on a raspberry pi, for instance, we often want to reduce write load of the sd card, then, the wal directory is mounted on a memory (space limited) partition.

the default value of the segment max file size, pushed the size of directory to 128 MB for each segment , which is too much ram consumption on a rasp.

the initial discussion is at https://github.com/prometheus/tsdb/pull/450

### CODE REVIEW

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -164,6 +164,10 @@ func main() {
 		"Maximum duration compacted blocks may span. For use in testing. (Defaults to 10% of the retention period).").
 		Hidden().PlaceHolder("<duration>").SetValue(&cfg.tsdb.MaxBlockDuration)
 
+	a.Flag("storage.tsdb.wal-segment-size",
+		"Maximum Size for tsdb Walk Ahead Log segment files").
+		Hidden().IntVar(&cfg.tsdb.WALSegmentSize)
```
Comment (krasi-georgiev) [inline cmd/prometheus/main.go]: I think we should accept this with an unit and parse it.
100KB,100MB,10GB

**Comment by brian-brazil** [comment]
Is this really something we should allow a user to configure? I can see users messing around with it (like they do with min/max block sizes) and just causing themselves problems.

**Comment by krasi-georgiev** [comment]
I think if we add some sane restrictions for min and max it should be fine and if not can always revert it at some point.

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -164,6 +164,10 @@ func main() {
 		"Maximum duration compacted blocks may span. For use in testing. (Defaults to 10% of the retention period).").
 		Hidden().PlaceHolder("<duration>").SetValue(&cfg.tsdb.MaxBlockDuration)
 
+	a.Flag("storage.tsdb.wal-segment-size",
+		"Maximum Size for tsdb Walk Ahead Log segment files").
```
Comment (krasi-georgiev) [inline cmd/prometheus/main.go]:
```suggestion
		"Size at which to split the tsdb WAL segment files(e.g. 10MB,100MB)").
```

**Comment by krasi-georgiev** [comment]
@glutamatt can you add some limits. Maybe: min 10mb and max 256mb

**Comment by krasi-georgiev** [comment]
also can you check if we can add a test to ensure that setting this flag creates the correct cfg.tsdb options.

**Comment by glutamatt** [comment]
> also can you check if we can add a test to ensure that setting this flag creates the correct cfg.tsdb options.

the only way i found of checking this is to :
1. dispatch a startup log about  WALSegmentSize
2. test in running a prometheus main command with a fixed flag value
3. check the logs of the runned command to verify if the final value in bytes is printed in logs

i'm trying to code this but i'm not sure it's the right way of testing it

**Comment by krasi-georgiev** [comment]
yeah doesn't sound a good way to test it. 
Just add the min/max limits and that should do it for now.

File: storage/tsdb/tsdb.go [hunk 3]
```diff
@@ -117,6 +121,9 @@ type Options struct {
 	// The maximum timestamp range of compacted blocks.
 	MaxBlockDuration model.Duration
 
+	// The maximum size of each WAL segment files
```
Comment (juliusv) [inline storage/tsdb/tsdb.go]: "files" -> "file"

nit: add period at end of comment.

File: cmd/prometheus/main.go [hunk 4]
```diff
@@ -164,6 +164,10 @@ func main() {
 		"Maximum duration compacted blocks may span. For use in testing. (Defaults to 10% of the retention period).").
 		Hidden().PlaceHolder("<duration>").SetValue(&cfg.tsdb.MaxBlockDuration)
 
+	a.Flag("storage.tsdb.wal-segment-size",
+		"Size at which to split the tsdb WAL segment files(e.g. 100MB)").
```
Comment (juliusv) [inline cmd/prometheus/main.go]: nit: add space before opening parenthesis.

**Comment by glutamatt** [comment]
> Just add the min/max limits and that should do it for now.

Done

File: cmd/prometheus/main.go (hunk 4 above)
Comment (glutamatt) [reply cmd/prometheus/main.go]: Done

File: storage/tsdb/tsdb.go (hunk 3 above)
Comment (glutamatt) [reply storage/tsdb/tsdb.go]: Done

File: storage/tsdb/tsdb.go [hunk 5]
```diff
@@ -180,8 +185,16 @@ func Open(path string, l log.Logger, r prometheus.Registerer, opts *Options) (*t
 		}
 	}
 
+	walSegmentSize := int(opts.WALSegmentSize)
+	if walSegmentSize != 0 {
+		if walSegmentSize < 10*1024*1024 || walSegmentSize > 256*1024*1024 {
+			return nil, errors.New("tsdb.Options.WALSegmentSize must be set between 10MB and 256MB")
```
Comment (krasi-georgiev) [inline storage/tsdb/tsdb.go]: maybe best to have this after this line.
https://github.com/prometheus/prometheus/blob/master/cmd/prometheus/main.go#L562

Comment (glutamatt) [reply storage/tsdb/tsdb.go]: yep i agress, done

Comment (glutamatt) [reply storage/tsdb/tsdb.go]: Maybe i could add a test to check if the app stop the right way if the flag value is set out of authorized bounds ?

Comment (krasi-georgiev) [reply storage/tsdb/tsdb.go]: yeah that is not a bad idea.

Comment (glutamatt) [reply storage/tsdb/tsdb.go]: Done !

**Comment by krasi-georgiev** [comment]
LGTM

File: storage/tsdb/tsdb.go [hunk 6]
```diff
@@ -19,6 +19,8 @@ import (
 	"time"
 	"unsafe"
 
+	"github.com/alecthomas/units"
+
```
Comment (simonpasquier) [inline storage/tsdb/tsdb.go]: The blank line is not needed.

File: storage/tsdb/tsdb.go [hunk 7]
```diff
@@ -117,6 +119,9 @@ type Options struct {
 	// The maximum timestamp range of compacted blocks.
 	MaxBlockDuration model.Duration
 
+	// The maximum size of each WAL segment file.
+	WALSegmentSize units.Base2Bytes
```
Comment (simonpasquier) [inline storage/tsdb/tsdb.go]: I'd rather use `int` here.

File: storage/tsdb/tsdb.go (hunk 6 above)
Comment (glutamatt) [reply storage/tsdb/tsdb.go]: done thanks

File: storage/tsdb/tsdb.go (hunk 7 above)
Comment (glutamatt) [reply storage/tsdb/tsdb.go]: this is what i used before allowing the flag value to be a byte unit value thanks to https://github.com/alecthomas/units package. Then, i need WALSegmentSize to be the type units.Base2Bytes to make the parser works as expected

Comment (glutamatt) [reply storage/tsdb/tsdb.go]: this is the doc of https://godoc.org/github.com/alecthomas/kingpin#ArgClause.BytesVar i used here https://github.com/prometheus/prometheus/pull/5029/files#diff-fc1d26df0adf6edc1b9f25c878ab9ccaR169

Comment (simonpasquier) [reply storage/tsdb/tsdb.go]: Ah right I missed that `Options` was used directly in the main func.

**Pull request merged**